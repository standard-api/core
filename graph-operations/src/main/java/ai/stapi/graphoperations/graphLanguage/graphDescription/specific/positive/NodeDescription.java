package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.ArrayList;
import java.util.List;

public class NodeDescription extends AbstractNodeDescription {

  public static final String SERIALIZATION_TYPE = "NodeDescription";

  protected NodeDescription() {

  }

  public NodeDescription(
      NodeDescriptionParameters parameters,
      List<GraphDescription> childDeclarations
  ) {
    super(SERIALIZATION_TYPE, childDeclarations);
    this.parameters = parameters;
  }

  public NodeDescription(NodeDescriptionParameters parameters) {
    super(SERIALIZATION_TYPE, new ArrayList<>());
    this.parameters = parameters;
  }

  public NodeDescription(
      NodeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
  }

  public NodeDescription(
      String nodeType,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, new NodeDescriptionParameters(nodeType), childDeclarations);
  }
}
