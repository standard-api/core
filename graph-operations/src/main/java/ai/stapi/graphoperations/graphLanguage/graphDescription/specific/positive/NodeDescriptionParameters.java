package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class NodeDescriptionParameters implements GraphDescriptionParameters {

  private String nodeType;

  private NodeDescriptionParameters() {

  }

  public NodeDescriptionParameters(String nodeType) {
    this.nodeType = nodeType;
  }

  public String getNodeType() {
    return nodeType;
  }
}