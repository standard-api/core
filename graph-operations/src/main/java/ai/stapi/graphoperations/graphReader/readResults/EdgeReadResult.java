package ai.stapi.graphoperations.graphReader.readResults;

import ai.stapi.graph.traversableGraphElements.TraversableEdge;

public class EdgeReadResult extends AbstractGraphElementReadResult {

  public EdgeReadResult(TraversableEdge edge) {
    super(edge);
  }

  public TraversableEdge getEdge() {
    return (TraversableEdge) this.getGraphElement();
  }

}
