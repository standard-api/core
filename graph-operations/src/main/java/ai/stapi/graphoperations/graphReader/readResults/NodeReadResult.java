package ai.stapi.graphoperations.graphReader.readResults;

import ai.stapi.graph.traversableGraphElements.TraversableNode;

public class NodeReadResult extends AbstractGraphElementReadResult {

  public NodeReadResult(TraversableNode node) {
    super(node);
  }

  public TraversableNode getNode() {
    return (TraversableNode) this.getGraphElement();
  }
}
