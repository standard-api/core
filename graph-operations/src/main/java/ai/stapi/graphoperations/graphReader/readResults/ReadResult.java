package ai.stapi.graphoperations.graphReader.readResults;

public interface ReadResult {

}
