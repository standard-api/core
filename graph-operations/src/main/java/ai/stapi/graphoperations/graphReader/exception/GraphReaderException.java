package ai.stapi.graphoperations.graphReader.exception;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.removal.RemovalGraphDescription;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.graphReader.readResults.ReadResult;
import ai.stapi.graphoperations.graphReader.readResults.ValueReadResult;
import ai.stapi.identity.UniqueIdentifier;
import org.apache.commons.lang3.StringUtils;

public class GraphReaderException extends RuntimeException {

  private GraphReaderException(String message) {
    super(message);
  }

  public static GraphReaderException becauseStartingElementWasNotFoundInProvidedGraph(
      UniqueIdentifier startElementId,
      String startElementType,
      String startElementGraphType
  ) {
    return new GraphReaderException(
        StringUtils.capitalize(startElementGraphType) + " '"
            + startElementId.toString()
            + "' of type '"
            + startElementType
            + "' not found in provided graph structure."
    );
  }

  public static GraphReaderException becauseStartingElementWasNotFoundInProvidedGraph(
      UniqueIdentifier startElementId) {
    return new GraphReaderException(
        "Graph element with ID ["
            + startElementId.toString()
            + "] was not found in provided graph structure."
    );
  }

  public static GraphReaderException becauseThereIsNoSupportingResolverForMappingPart(
      PositiveGraphDescription graphDescription) {
    return new GraphReaderException(
        "There is no supporting resolver for Graph Description '" + graphDescription.getClass()
            .getSimpleName() + "'."
    );
  }

  public static GraphReaderException becauseThereIsMultipleSupportingResolverForMappingPart(
      PositiveGraphDescription graphDescription) {
    return new GraphReaderException(
        "There is multiple supporting resolvers for Graph Description '"
            + graphDescription.getClass().getSimpleName() + "'."
    );
  }

  public static GraphReaderException becauseGraphDescriptionContainsRemovalGraphDescription(
      GraphDescription description) {
    return new GraphReaderException(
        "Provided "
            + GraphDescription.class.getSimpleName()
            + " contains " + RemovalGraphDescription.class.getSimpleName()
            + " of type '" + description.getClass().getSimpleName()
            + "' and that is not allowed in " + GraphReaderException.class.getSimpleName()
            + "."
    );
  }

  public static GraphReaderException becauseMappingPartReturnsTypeIncompatibleWithNext(
      ObjectGraphMapping currentMappingPart,
      ObjectGraphMapping nextMappingPart
  ) {
    return new GraphReaderException(
        "Mapping part '"
            + currentMappingPart.getClass().getSimpleName()
            + "' returns '"
            + currentMappingPart.getSerializationType()
            + "' ,but following mapping part '"
            + nextMappingPart.getClass().getSimpleName()
            + "' expects '"
            + nextMappingPart.getSerializationType() + "'."
    );
  }

  public static GraphReaderException becauseProvidedGraphDescriptionCanNotBeFirst(
      PositiveGraphDescription graphDescription) {
    return new GraphReaderException(
        "Provided Graph Description of type '"
            + graphDescription.getClass().getSimpleName()
            + "' can't be as first in the composite."
    );
  }

  public static GraphReaderException becauseReadResultsFromLastMappingPartIsEmpty(
      ObjectGraphMapping mappingPart, Integer index) {
    return new GraphReaderException(
        "Read Results from mapping part '"
            + mappingPart.getClass().getSimpleName()
            + "' at index '"
            + index.toString()
            + "' was empty. That is not allowed when reading."
    );
  }

  public static GraphReaderException becauseGraphDescriptionHasMultipleChildren(
      PositiveGraphDescription graphDescription) {
    return new GraphReaderException(
        "Graph Description of type '"
            + graphDescription.getClass().getSimpleName()
            + "' contains multiple children and that is not allowed in Graph Reader."
    );
  }

  public static GraphReaderException becauseLastReadResultsAreNotValues(ReadResult readResult) {
    return new GraphReaderException(
        "When reading value, last Read Result must be of type '"
            + ValueReadResult.class.getSimpleName()
            + "' but was of type '"
            + readResult.getClass().getSimpleName()
            + "' instead."
    );
  }
}
