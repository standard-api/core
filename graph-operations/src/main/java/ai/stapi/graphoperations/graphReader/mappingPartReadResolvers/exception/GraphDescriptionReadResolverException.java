package ai.stapi.graphoperations.graphReader.mappingPartReadResolvers.exception;

import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.graphoperations.graphReader.readResults.ReadResult;
import ai.stapi.graphoperations.graphReader.mappingPartReadResolvers.GraphDescriptionReadResolver;
import ai.stapi.identity.UniqueIdentifier;

public class GraphDescriptionReadResolverException extends RuntimeException {

  private GraphDescriptionReadResolverException(String message) {
    super(message);
  }

  public static GraphDescriptionReadResolverException becauseIngoingReadResultIsNotSupported(
      ReadResult readResult,
      GraphDescriptionReadResolver resolver
  ) {
    return new GraphDescriptionReadResolverException(
        "Provided previous Read Result of type '"
            + readResult.getClass().getSimpleName()
            + "' is not supported by resolver '"
            + resolver.getClass().getSimpleName()
            + "'."
    );
  }

  public static GraphDescriptionReadResolverException becauseEdgeDoesNotContainNodeOfGivenType(
      UniqueIdentifier edgeId, String nodeType) {
    return new GraphDescriptionReadResolverException(
        "Edge '" + edgeId.toString() + "' does not contain node of type '" + nodeType + "'."
    );
  }

  public static GraphDescriptionReadResolverException becauseGraphDescriptionTypeIsNotSupported(
      PositiveGraphDescription graphDescription,
      GraphDescriptionReadResolver resolver
  ) {
    return new GraphDescriptionReadResolverException(
        "Graph Description of type '"
            + graphDescription.getClass().getSimpleName()
            + "' is not supported by '"
            + resolver.getClass().getSimpleName()
            + "' resolver."
    );
  }

  public static GraphDescriptionReadResolverException becauseAttributeDoesNotExists(
      UniqueIdentifier graphElementId,
      String attributeName
  ) {
    return new GraphDescriptionReadResolverException(
        "Graph Element with id '" + graphElementId + "' does not contain attribute '"
            + attributeName + "'."
    );
  }

}
