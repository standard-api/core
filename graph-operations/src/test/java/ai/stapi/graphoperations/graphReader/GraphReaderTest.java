package ai.stapi.graphoperations.graphReader;

import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graphoperations.graphLanguage.graphDescription.graphDescriptionBuilder.GraphDescriptionBuilder;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graph.Graph;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graphoperations.graphReader.ExampleImplementation.ExampleAmbiguousGraphDescription;
import ai.stapi.graphoperations.graphReader.ExampleImplementation.ExampleNotSupportedGraphDescription;
import ai.stapi.graphoperations.graphReader.exception.GraphReaderException;
import ai.stapi.graphoperations.graphReader.mappingPartReadResolvers.exception.GraphDescriptionReadResolverException;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.graphoperations.graphReader.readResults.EdgeReadResult;
import ai.stapi.graphoperations.graphReader.readResults.NodeReadResult;
import ai.stapi.graphoperations.graphReader.readResults.ReadResult;
import ai.stapi.test.integration.IntegrationTestCase;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;

class GraphReaderTest extends IntegrationTestCase {

  @Autowired
  private GraphReader graphReader;

  @Test
  void itThrowsErrorWhenFirstGraphElementIsNotFound() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var exampleGraph = this.getSimpleExampleGraph(nodeId).traversable();
    Executable throwable = () -> this.graphReader.read(
        UniversallyUniqueIdentifier.fromString("3db53fc1-4e24-406c-8a88-fbfc71e1b43c"),
        new NodeDescription(new NodeDescriptionParameters("house")),
        exampleGraph
    );
    this.thenExceptionMessageApproved(GraphReaderException.class, throwable);
  }

  @Test
  void itThrowsWhenThereIsNoSupportingResolverForMappingParts() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var exampleGraph = this.getSimpleExampleGraph(nodeId).traversable();
    Executable throwable = () -> this.graphReader.read(
        nodeId,
        new NodeDescription(new NodeDescriptionParameters("house"),
            new ExampleNotSupportedGraphDescription()),
        exampleGraph
    );
    this.thenExceptionMessageApproved(GraphReaderException.class, throwable);
  }

  @Test
  void itThrowsWhenThereIsMultipleSupportingResolversForMappingParts() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var exampleGraph = this.getSimpleExampleGraph(nodeId).traversable();
    Executable throwable = () -> this.graphReader.read(
        nodeId,
        new NodeDescription(new NodeDescriptionParameters("house"),
            new ExampleAmbiguousGraphDescription()),
        exampleGraph
    );
    this.thenExceptionMessageApproved(GraphReaderException.class, throwable);
  }

  @Test
  void itThrowsErrorWhenThereAreIncompatibleInstructionsAfterEachOther() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("house")
        .addNodeDescription("window");
    var exampleGraph = this.getSimpleExampleGraph(nodeId).traversable();
    Executable throwable = () -> this.graphReader.read(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        exampleGraph
    );
    this.thenExceptionMessageApproved(GraphDescriptionReadResolverException.class, throwable);
  }

  @Test
  void itCanReadNode() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("house")
        .addOutgoingEdge("has_doors")
        .addNodeDescription("doors")
        .addOutgoingEdge("has_doors_handle")
        .addNodeDescription("doors_handle");
    var graph = this.getSimpleExampleGraph(nodeId).traversable();
    var result = this.graphReader.read(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenApproveReadResult(result);
  }

  @Test
  void itCanReadNodes() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("country")
        .addOutgoingEdge("has_region")
        .addNodeDescription("region")
        .addOutgoingEdge("has_city")
        .addNodeDescription("city");
    var graph = this.getComplexExampleGraph(nodeId).traversable();
    var result = this.graphReader.read(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenApproveReadResult(result);
  }

  @Test
  void itCanReadNodeAttribute() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("house")
        .addLeafAttribute("material")
        .addStringAttributeValue();
    var graph = this.getSimpleExampleGraph(nodeId).traversable();
    var result = this.graphReader.<String>readValues(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenObjectApproved(result);
  }

  @Test
  void itCanReadNodesAttributes() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("country")
        .addOutgoingEdge("has_region")
        .addNodeDescription("region")
        .addLeafAttribute("name")
        .addStringAttributeValue();
    var graph = this.getComplexExampleGraph(nodeId).traversable();
    var result = this.graphReader.readValues(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenObjectApproved(result.stream().sorted().toList());
  }

  @Test
  void itWillReturnEmptyListWhenAttributeNotPresent() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("house")
        .addLeafAttribute("nonExistent")
        .addStringAttributeValue();
    var graph = this.getSimpleExampleGraph(nodeId).traversable();
    var result = this.graphReader.<String>readValues(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenObjectApproved(result);
  }
  
  @Test
  void itThrowsErrorWhenThereIsMultipleBranches() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    var nodeBranch = builder.addNodeDescription("country")
        .addOutgoingEdge("has_region")
        .addNodeDescription("region");
    nodeBranch.addOutgoingEdge("random_edge")
        .addLeafAttribute("name")
        .addStringAttributeValue();
    nodeBranch
        .addLeafAttribute("random_attribute")
        .addStringAttributeValue();
    var graph = this.getComplexExampleGraph(nodeId).traversable();
    Executable throwable = () -> this.graphReader.readValues(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenExceptionMessageApproved(GraphReaderException.class, throwable);
  }

  @Test
  void itCanReadEdge() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("house")
        .addOutgoingEdge("has_window");
    var graph = this.getSimpleExampleGraph(nodeId).traversable();
    var result = this.graphReader.read(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenApproveReadResult(result);
  }

  @Test
  void itCanReadEdges() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("country")
        .addOutgoingEdge("has_region")
        .addNodeDescription("region")
        .addOutgoingEdge("has_city");
    var graph = this.getComplexExampleGraph(nodeId).traversable();
    var result = this.graphReader.read(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenApproveReadResult(result);
  }

  @Test
  void itCanReadEdgeAttribute() {
    var nodeId = UniversallyUniqueIdentifier.randomUUID();
    var graph = this.getSimpleExampleGraph(nodeId).traversable();
    var builder = new GraphDescriptionBuilder();
    builder.addNodeDescription("house")
        .addOutgoingEdge("has_doors")
        .addLeafAttribute("position")
        .addStringAttributeValue();
    var result = this.graphReader.<String>readValues(
        nodeId,
        builder.getOnlyPositiveGraphDescriptions(),
        graph
    );
    this.thenObjectApproved(result);
  }

  private Graph getSimpleExampleGraph(UniqueIdentifier startNodeId) {
    var graph = new Graph();
    var houseNode = new Node(startNodeId, "house");
    houseNode =
        houseNode.add(new LeafAttribute<>("material", new StringAttributeValue("bricks")));
    var windowNode = new Node("window");
    windowNode =
        windowNode.add(new LeafAttribute<>("material", new StringAttributeValue("glass")));
    var doorsNode = new Node("doors");
    doorsNode =
        doorsNode.add(new LeafAttribute<>("material", new StringAttributeValue("wood")));
    var doorsHandleNode = new Node("doors_handle");
    doorsHandleNode = doorsHandleNode.add(
        new LeafAttribute<>("material", new StringAttributeValue("metal")));
    var doorEdge = new Edge(houseNode, "has_doors", doorsNode);
    doorEdge = doorEdge.add(new LeafAttribute<>("position", new StringAttributeValue("front")));
    graph = graph.withAll(houseNode, windowNode, doorsNode, doorsHandleNode);
    graph = graph.withAll(
        doorEdge,
        new Edge(houseNode, "has_window", windowNode),
        new Edge(doorsNode, "has_doors_handle", doorsHandleNode)
    );
    return graph;
  }

  private Graph getComplexExampleGraph(UniqueIdentifier startNodeId) {
    var graph = new Graph();
    var countryNode = new Node(startNodeId, "country");
    countryNode =
        countryNode.add(new LeafAttribute<>("name", new StringAttributeValue("Czechia")));
    var cityNode = new Node("city");
    cityNode = cityNode.add(new LeafAttribute<>("name", new StringAttributeValue("Prague")));
    var cityNode1 = new Node("city");
    cityNode1 = cityNode1.add(new LeafAttribute<>("name", new StringAttributeValue("Brno")));
    var cityNode2 = new Node("city");
    cityNode2 = cityNode2.add(new LeafAttribute<>("name", new StringAttributeValue("Plzen")));
    var regionNode = new Node("region");
    regionNode =
        regionNode.add(new LeafAttribute<>("name", new StringAttributeValue("Cechy")));
    var regionNode1 = new Node("region");
    regionNode1 =
        regionNode1.add(new LeafAttribute<>("name", new StringAttributeValue("Morava")));
    graph = graph.withAll(countryNode, regionNode, regionNode1, cityNode, cityNode1, cityNode2);
    graph = graph.withAll(
        new Edge(countryNode, "has_region", regionNode),
        new Edge(countryNode, "has_region", regionNode1),
        new Edge(regionNode, "has_city", cityNode),
        new Edge(regionNode1, "has_city", cityNode1),
        new Edge(regionNode, "has_city", cityNode2)
    );
    return graph;
  }

  protected void thenApproveReadResult(List<ReadResult> result) {

    if (result.get(0) instanceof NodeReadResult) {
      var list = result.stream()
          .map(readResult -> {
            var nodeResult = (NodeReadResult) readResult;
            return nodeResult.getNode();
          })
          .toList();
      this.thenNodesApproved(list);
    }

    if (result.get(0) instanceof EdgeReadResult) {
      var list = result.stream()
          .map(readResult -> {
            var edgeResult = (EdgeReadResult) readResult;
            return edgeResult.getEdge();
          })
          .toList();
      this.thenEdgesApproved(list);
    }
  }
}