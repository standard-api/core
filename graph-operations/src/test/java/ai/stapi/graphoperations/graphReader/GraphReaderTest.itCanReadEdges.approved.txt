edge_type: has_city
edge_hash: 36F4F
region -> has_city -> city (BF5F2 -> 36F4F -> 6FA63)
attributes:

edge_type: has_city
edge_hash: 6F39C
region -> has_city -> city (B510A -> 6F39C -> B268D)
attributes:

edge_type: has_city
edge_hash: C7887
region -> has_city -> city (B510A -> C7887 -> F842E)
attributes:

